akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala

```diff
@@ -38,6 +38,23 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
 
   val http = Http(system)
 
+  def bindAndHandle(
+    handler:   Flow[HttpRequest, HttpResponse, Any],
+    interface: String, port: Int = DefaultPortForProtocol,
+    connectionContext: ConnectionContext = http.defaultServerHttpContext,
+    settings:          ServerSettings    = ServerSettings(system),
+    parallelism:       Int               = 1,
+    log:               LoggingAdapter    = system.log)(implicit fm: Materializer): Future[ServerBinding] = {
+    bindAndHandleImpl(
+      Http2Blueprint.handleWithStreamIdHeader(handler),
```

Comment:
I see how that is nice for existing, non-reordering handler flows.

However, it also loses a significant part of HTTP2's appeal because it reintroduces head-of-line blocking. It would still be able to multiplex response streams, yes, but any request handling that can only return a response with an appropriate status code after some delay (like having to wait for the result of async callbacks), will block other responses on the stream from being sent out (and even from being processed when blocked responses pile up).

In the end, the `bindAndHandle` API could offer two different semantics under the same signature which both have drawbacks for the user:
 * the one suggested here where the user flow must be ordered one-to-one (which matches the original HTTP/1.1 semantics with HTTP pipelining enabled)
 * the one offered by `bindAndHandleImpl` where there are no constraints on flow ordering but the flow has to ensure that the `streamId` attribute is manually propagated

The second one is the functionally more useful, so we should provide it in some way.

Reply:
Right - and the current `Route.toFlow` produces a non-parallel one-to-one `Flow`, so after that the 'damage is done' and we can't really untangle it anymore.

Does this mean we need a `bindAndHandle` signature that doesn't accept a `Flow`, but accepts something that gets turned into the right kind of `Flow` only after negotiation?

It kindof sounds like what we really want is indeed `bindAndHandleAsync`, perhaps with an utility for automatically converting the `Route` to `HttpRequest => Future[HttpResponse]` (we already have `Route.toFunction`). Of course the `bindAndHandle` that takes a `Flow` must stay, but then how do we prevent users from accidentally using that one with a `Route` that was meant to support Http2?

Removing the implicit `Route.toFlow` and instead making `Route.toFunction` implicit and providing a `bindAndHandle` that takes such a function might work, but seems like a pretty invasive change.

Reply:
Could `Route.toFunction` be implicit as well? Then, we could say, that you should use `bindAndHandleAsync` with Routes if you want to support both Http 1 and 2.

We could still fix `Route.toFlow` as well to propagate the attribute but using that you would still lose proper multiplexing with HTTP/2.

But I'm also open for the more drastic approach of deprecating `RouteResult.routeToFlow` pointing to using `bindAndHandleAsync` instead (or i.e. to change the deprecation warning of `route2handlerflow` and adapting the message to say that). After all, the high-level DSL is not stream-based and is unlikely to become stream-based in the future.

Reply:
(first step, allowing a `Route` in `bindAndHandleAsync`, is in https://github.com/akka/akka-http/pull/3237)

Reply:
:+1: for a `bind` overload if that works out.